xdsl/irdl/constraints.py

```diff
@@ -90,67 +97,10 @@ def update(self, other: ConstraintContext):
 """
 
 
-_T = TypeVar("_T")
-
-
-@dataclass(frozen=True)
-class VarExtractor(Generic[_T], abc.ABC):
-    """
-    A VarExtractor describes how to extract the value of a constraint variable
-    from a value of type _T. The result of the extraction must be a subtype of
-    ConstraintVariableType.
-    """
-
-    @abstractmethod
-    def extract_var(self, a: _T) -> ConstraintVariableType: ...
-
-    def flatten(self) -> Iterator[VarExtractor[_T]]:
-        """Helper function to merge extractors, should only be overloaded by MergeExtractor"""
-        yield self
-
-
-@dataclass(frozen=True)
-class MergeExtractor(VarExtractor[_T]):
-    variable: str
-    extractors: tuple[VarExtractor[_T], ...]
-    """Extractors to merge, should not be empty."""
-
-    def extract_var(self, a: _T) -> ConstraintVariableType:
-        res = set(x.extract_var(a) for x in self.extractors)
-        if len(res) == 1:
-            return res.pop()
-        else:
-            raise ValueError(
-                f"Value of variable {self.variable} could not be uniquely extracted.\n"
-                f"Possible values are: {{{', '.join(sorted(str(x) for x in res))}}}"
-            )
-
-    def flatten(self) -> Iterator[VarExtractor[_T]]:
-        return iter(self.extractors)
-
-
-def merge_extractors(variable: str, *extractors: VarExtractor[_T]) -> VarExtractor[_T]:
-    """
-    Helper method used for merging dictionaries of extractors.
-    Must not be called with no extractors.
-    """
-    if len(extractors) == 1:
-        return extractors[0]
-    extractor_tuple = tuple(x for ex in extractors for x in ex.flatten())
-    return MergeExtractor(variable, extractor_tuple)
-
-
-def merge_extractor_dicts(
-    *dicts: dict[str, VarExtractor[_T]],
-) -> dict[str, VarExtractor[_T]]:
-    """
-    Merge dictionaries of extractors.
-    The keys of the resulting dictionary is equal to the union of the keys
-    of the input dictionaries, with duplicated values being merged to a
-    single MergeExtractor
-    """
-    keys = set(key for d in dicts for key in d)
-    return {v: merge_extractors(v, *(d[v] for d in dicts if v in d)) for v in keys}
+class ConstraintVarType(StrEnum):
+    ATTRIBUTE = auto()
+    RANGE = auto()
+    INT = auto()
```

Comment:
Where is this used?

Reply:
It's not used in this at all. I started doing this because I wanted to be able to implement an `AttributeConstraintVariableDirective` which would allow you to put something like `%T` in assembly format for a constraint variable `T`. The problem is that there's no easy way to have an operation definition and variable name `T` and tell whether `T` is a attribute variable, range variable, or int variable, which you need to know in order to tell how to parse. That said I imagine this should be a separate PR.